art/attacks/evasion/pixel_threshold.py

```diff
@@ -0,0 +1,1457 @@
+                recombination=1,
+                atol=-1,
+                callback=callback_fn,
+                polish=False)
+            adv_x = strategy.x
+
+        if self._attack_success(adv_x, image, target_class):
+            return True, self._perturb_image(adv_x, image)[0]
+        else:
+            return False, image
+
+
+class PixelAttack(PixelThreshold):
+    """
+    This attack was originally implemented by Vargas et al. (2019).
+    It is generalisation of One Pixel Attack originally implemented by
+    Su et al. (2019)
+
+    | One Pixel Attack Paper link:
+        https://ieeexplore.ieee.org/abstract/document/8601309/citations#citations
+        (arXiv link: https://arxiv.org/pdf/1710.08864.pdf)
+    | Pixel Attack Paper link:
+        https://arxiv.org/abs/1906.06026
+    """
+
+    def __init__(
+            self,
+            classifier,
+            th=None,
+            es=0,
+            targeted=False,
+            verbose=False):
+        """
+        Create a :class:`.PixelAttack` instance.
+        """
+        super(
+            PixelAttack,
+            self).__init__(
+                classifier,
+                th,
+                es,
+                targeted,
+                verbose)
+        self.type_attack = 0
+
+    def _perturb_image(self, x, img):
+        """
+        Perturbs the given image `img` with the given perturbation `x`.
+        """
+        if x.ndim < 2:
+            x = np.array([x])
+        imgs = np.tile(img, [len(x)] + [1] * (x.ndim + 1))
+        x = x.astype(int)
+        for adv, image in zip(x, imgs):
+            for pixel in np.split(adv, len(adv) // (2 + image.shape[-1])):
+                x_pos, y_pos, *rgb = pixel
+                image[x_pos % image.shape[-3], y_pos % image.shape[-2]] = rgb
+        return imgs
+
+    def _get_bounds(self, img, limit):
+        """
+        Define the bounds for the image `img` within the limits `limit`.
+        """
+        initial = []
+        if self.es == 0:
+            for count, (i, j) in enumerate(
+                    product(range(img.shape[-3]), range(img.shape[-2]))):
+                initial += [i, j]
+                for k in range(img.shape[-1]):
+                    initial += [img[i, j, k]]
+                if count == limit - 1:
+                    break
+                else:
+                    continue
+            min_bounds = [0, 0]
+            for _ in range(img.shape[-1]):
+                min_bounds += [0]
+            min_bounds = min_bounds * limit
+            max_bounds = [img.shape[-3], img.shape[-2]]
+            for _ in range(img.shape[-1]):
+                max_bounds += [255]
+            max_bounds = max_bounds * limit
+            bounds = [min_bounds, max_bounds]
+        else:
+            bounds = [(0, img.shape[-3]), (0, img.shape[-2])]
+            for _ in range(img.shape[-1]):
+                bounds += [(0, 255)]
+            bounds = bounds * limit
+        return bounds, initial
```

Comment:
As requested, this is the first instance where input image shape is used. 
This function defines the bound limit for the pixel attack. 
Please note that the Pixel Attack perturbation is a vector of 5 variables for 3 channel image and 3 variable for 1 channel image, 
`For 3 Channel Index: <Position of X>, <Position of Y>, <Value of Red Channel>, <Value of Blue Channel>, <Value of Green Channel>`
`For 1Channel Index: <Position of X>, <Position of Y>, <Value of Monochrome Channel>'.

I tried to use `classifer.input_shape` but this parameter is not mandatory for all classifiers, therefore, I did not use it.

Reply:
If you would use `classifer.input_shape`, could you avoid `swapaxes`?

Reply:
Indeed, if I initially have `classifer.input_shape`, I could definitely avoid `swapaxes`.